### PR TITLE
JSON Help + Lib Functions adding

### PR DESCRIPTION
**Pull request by apeltzer**
As on the tin this adds:

- Automated help generation from JSON
- Summary generation from JSON
- Some extra functions (e.g. awsbatch checks, hostname checks) are callable from the external lib now

We could have the same as in rnaseq: 

- check conda channels for correctness (!)
- ...

### CODE REVIEW

**Comment by apeltzer** [comment]
Fixes #676

**Comment by apeltzer** [comment]
Ok, just switched over to the newest lib adding in functionality from here https://github.com/nf-core/tools/pull/852 

Note, that I have NOT changed how we do the help + summary, as the newest nf-core template will NOT contain that most likely for now. I've intentionally kept that in the `Schema.groovy` file and the main validation stuff in the `NfcoreSchema.groovy` part that will be kept up2date via nf-core sync later. This will make our lifes easier hopefully ;-) 

Schema.groovy:

- help
- summary
- multiqc summary stuff

Headers.groovy

- Header ;-) 

NfcoreSchema.groovy

- Validation

**Comment by ewels** [comment]
So fast, we haven't even merged this to tools yet 😆

**Comment by apeltzer** [comment]
> So fast, we haven't even merged this to tools yet 😆

I'm just too excited about this 🎉

File: lib/Checks.groovy [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+               "* The nf-core framework\n" +
+               "  https://dx.doi.org/10.1038/s41587-020-0439-x\n" +
+               "  https://rdcu.be/b1GjZ\n\n" +
+               "* Software dependencies\n" +
+               "  https://github.com/${workflow.manifest.name}/blob/master/CITATIONS.md"
+    }
+
+    // Exit pipeline if incorrect --genome key provided
+    static void genome_exists(params, log) {
+        if (params.genomes && params.genome && !params.genomes.containsKey(params.genome)) {
```
Comment (jfy133) [inline lib/Checks.groovy:81]:
```suggestion
        if (params.genome && !params.genomes.containsKey(params.genome)) {
```

File: lib/Checks.groovy [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    // Print a warning if using '--aligner star_rsem' and '--with_umi'
+    static void rsem_umi_error(log) {
+        log.error "=============================================================================\n" +
+                  "  When using '--aligner star_rsem', STAR is run by RSEM itself and so it is\n" +
+                  "  not possible to remove UMIs before the quantification.\n\n" +
+                  "  If you would like to remove UMI barcodes using the '--with_umi' option\n" + 
+                  "  please use either '--aligner star_salmon' or '--aligner hisat2'.\n" +
+                  "============================================================================="
+        System.exit(0)
+    }
```
Comment (jfy133) [inline lib/Checks.groovy:163]: this seems specific to rnaseq, no?

EDIT: and stuff below?

File: main.nf [hunk 3]
```diff
@@ -12,251 +12,27 @@
...
+    log.info NfcoreSchema.params_help(workflow, params, json_schema, command)
     exit 0
 }
 
+////////////////////////////////////////////////////
+/* --         VALIDATE PARAMETERS              -- */
+////////////////////////////////////////////////////
+
+def unexpectedParams = []
+if (params.validate_params) {
```
Comment (jfy133) [inline main.nf:29]: Is this param something one has to manually add at CLI execution?

EDIT: Ah ok, set to true by default :+1:

File: nextflow_schema.json [hunk 4]
```diff
@@ -587,7 +605,6 @@
                 "bt2n": {
                     "type": "integer",
                     "description": "Specify the -N parameter for bowtie2 (mismatches in seed). This will override defaults from alignmode/sensitivity.",
-                    "default": 0,
```
Comment (jfy133) [inline nextflow_schema.json:590]: Why has this been removed (and below)?

Comment (apeltzer) [reply nextflow_schema.json:590]: I fear that was when updating with the schema builder...?

File: lib/Checks.groovy (hunk 1 above)
Comment (apeltzer) [reply lib/Checks.groovy]: I removed it for now as we won't have that in the template / generic version for now - can do that later

File: nextflow_schema.json (hunk 4 above)
Comment (apeltzer) [reply nextflow_schema.json:590]: Old version of tools i suppose (1.12.X?)

**Review by jfy133 — APPROVED** [review]

**Pull request merged**